### PR TITLE
ci: gate unwind tests and add ci step for tesing with panic=abort

### LINKED ISSUE #6109: Verify that all tests pass with panic=abort
We should make sure that the test suite passes with panic=abort. Tests that need unwinding can be gated with `#[cfg(panic = "unwind")]`.

To close this issue, add a new CI step that runs the test suite with panic=abort.

### PR DESCRIPTION
**Pull request by maminrayej**
Verifies that all tests pass with panic=abort by adding a step to CI. Also, gates tests that need unwinding with `#[cfg(panic = "unwind")]`.

Resolves #6109.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -134,6 +134,10 @@ jobs:
         uses: dtolnay/rust-toolchain@stable
         with:
             toolchain: ${{ env.rust_stable }}
+      - name: Install Rust ${{ env.rust_nightly }}
+        uses: dtolnay/rust-toolchain@stable
+        with:
+            toolchain: ${{ env.rust_nightly }}
```
Comment (Darksonn) [inline .github/workflows/ci.yml:140]: Please don't install multiple Rust versions in a single CI run. It causes confusion about which version of Rust is actually being used in each call. Instead, create a new CI run for this check.

File: tokio/tests/io_panic.rs
Comment (Darksonn) [inline tokio/tests/io_panic.rs:1]: For the tests with `_panic.rs` in the name, I'd just add `#![cfg(panic = "unwind")]` at the top and disable all tests in the file.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (maminrayej) [reply .github/workflows/ci.yml]: Okay. Created a new CI run.

File: tokio/tests/io_panic.rs
Comment (maminrayej) [reply tokio/tests/io_panic.rs:1]: You're right. Edited those files as suggested.

File: tokio/tests/io_panic.rs [hunk 2]
```diff
@@ -70,7 +71,6 @@ fn read_buf_initialize_unfilled_to_panic_caller() -> Result<(), Box<dyn Error>>
     Ok(())
 }
 
-#[cfg(panic = "unwind")]
```
Comment (taiki-e) [inline tokio/tests/io_panic.rs:73]: Could you remove #[cfg(panic = "unwind")] of read_buf_initialize_unfilled_to_panic_caller (line 57) as well?

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -148,6 +148,34 @@ jobs:
...
+        uses: taiki-e/install-action@v2
+        with:
+          tool: cargo-nextest
+
+      - uses: Swatinem/rust-cache@v2
+
+      - name: test all --all-features panic=abort
+        run: |
+          set -euxo pipefail
+          RUSTFLAGS="-C panic=abort -Zpanic-abort-tests" cargo nextest run --workspace --exclude tokio-macros --exclude tests-build --all-features --tests
```
Comment (taiki-e) [inline .github/workflows/ci.yml:177]:
```suggestion
          RUSTFLAGS="$RUSTFLAGS -C panic=abort -Zpanic-abort-tests" cargo nextest run --workspace --exclude tokio-macros --exclude tests-build --all-features --tests
```

**Review by taiki-e — COMMENTED** [review]
There appear to be a lot of unused_imports/dead_code warnings.
We may want to use `#[cfg_attr(panic = "abort", ignore)]` instead of using `#[cfg(panic = "unwind")]` on the function.

File: tokio/tests/io_panic.rs (hunk 2 above)
Comment (maminrayej) [reply tokio/tests/io_panic.rs:73]: Done.

**Review by taiki-e — APPROVED** [review]

**Pull request merged**